File: mobile-ios/ReactivePipeline.swift
import Foundation
import ReactiveSwift

struct SyncResult: Codable {
    let summary: Summary
    let conflicts: [SyncConflict]

    struct Summary: Codable {
        let appliedChanges: Int
        let pendingChanges: Int
        let lastSyncedAt: Date
    }
}

struct SyncConflict: Codable {
    let entity: String
    let entityId: String
    let reason: String
}

enum SyncError: Error {
    case transport(underlying: Error)
    case decoding
    case persistence(details: String)
}

protocol SyncNetworking {
    func requestSync(payload: Data) -> SignalProducer<Data, SyncError>
}

protocol SyncPersistence {
    func apply(snapshot: Data) -> SignalProducer<Void, SyncError>
}

protocol DomainMapper {
    func domainModels() -> SignalProducer<[JobViewModel], Never>
}

struct JobViewModel {
    let id: String
    let customerName: String
    let status: String
    let scheduledStart: Date?
    let scheduledEnd: Date?
}

final class FieldPrimePipeline {
    private let networking: SyncNetworking
    private let persistence: SyncPersistence
    private let mapper: DomainMapper

    init(networking: SyncNetworking, persistence: SyncPersistence, mapper: DomainMapper) {
        self.networking = networking
        self.persistence = persistence
        self.mapper = mapper
    }

    func sync(payload: Data) -> SignalProducer<SyncResult, SyncError> {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601

        return networking.requestSync(payload: payload)
            .start(on: QueueScheduler(qos: .userInitiated, name: "app.fieldprime.sync"))
            .attemptMap { data -> Result<SyncResult, SyncError> in
                guard let result = try? decoder.decode(SyncResult.self, from: data) else {
                    return .failure(.decoding)
                }
                return .success(result)
            }
            .flatMap(.concat) { [persistence] result -> SignalProducer<SyncResult, SyncError> in
                guard let snapshot = try? encoder.encode(result) else {
                    return SignalProducer(error: .decoding)
                }

                return persistence.apply(snapshot: snapshot)
                    .then(SignalProducer(value: result))
            }
            .observe(on: QueueScheduler(qos: .background, name: "app.fieldprime.persistence"))
    }

    func jobsStream() -> SignalProducer<[JobViewModel], Never> {
        return mapper.domainModels()
            .observe(on: QueueScheduler.main)
    }
}

File: mobile-android/ReactivePipeline.kt
package app.fieldprime.pipeline

import java.time.Instant
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.flatMapConcat
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.flow.map
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class SyncResult(
    val summary: Summary,
    val conflicts: List<SyncConflict>
) {
    @Serializable
    data class Summary(
        val appliedChanges: Int,
        val pendingChanges: Int,
        val lastSyncedAt: String
    )
}

@Serializable
data class SyncConflict(
    val entity: String,
    val entityId: String,
    val reason: String
)

sealed interface SyncError {
    data class Transport(val cause: Throwable) : SyncError
    data object Decoding : SyncError
    data class Persistence(val details: String) : SyncError
}

class SyncException(val reason: SyncError) : Exception(reason.toString())

interface SyncNetworking {
    fun requestSync(payload: ByteArray): Flow<Result<ByteArray>>
}

interface SyncPersistence {
    fun apply(snapshot: ByteArray): Flow<Result<Unit>>
}

interface DomainMapper {
    fun domainModels(): Flow<List<JobViewModel>>
}

data class JobViewModel(
    val id: String,
    val customerName: String,
    val status: String,
    val scheduledStart: Instant?,
    val scheduledEnd: Instant?
)

class FieldPrimePipeline(
    private val networking: SyncNetworking,
    private val persistence: SyncPersistence,
    private val mapper: DomainMapper,
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
) {
    fun sync(payload: ByteArray): Flow<Result<SyncResult>> {
        return networking.requestSync(payload)
            .flatMapConcat { response ->
                response.fold(
                    onSuccess = { body ->
                        deserializeSyncResult(body).fold(
                            onSuccess = { result ->
                                persistence.apply(serialize(result))
                                    .map { persistenceOutcome ->
                                        persistenceOutcome.fold(
                                            onSuccess = { Result.success(result) },
                                            onFailure = { Result.failure(mapPersistenceError(it)) }
                                        )
                                    }
                            },
                            onFailure = { error -> flow { emit(Result.failure(error)) } }
                        )
                    },
                    onFailure = { cause -> flow { emit(Result.failure(mapTransportError(cause))) } }
                )
            }
            .catch { emit(Result.failure(mapTransportError(it))) }
            .flowOn(ioDispatcher)
    }

    fun jobsStream(): Flow<List<JobViewModel>> {
        return mapper.domainModels()
            .flowOn(Dispatchers.Default)
    }

    private fun deserializeSyncResult(bytes: ByteArray): Result<SyncResult> =
        runCatching {
            Json.decodeFromString(SyncResult.serializer(), bytes.decodeToString())
        }.recoverCatching {
            throw SyncException(SyncError.Decoding)
        }

    private fun serialize(result: SyncResult): ByteArray {
        return Json.encodeToString(SyncResult.serializer(), result).encodeToByteArray()
    }

    private fun mapTransportError(cause: Throwable): Throwable = SyncException(SyncError.Transport(cause))

    private fun mapPersistenceError(cause: Throwable): Throwable =
        SyncException(
            SyncError.Persistence(details = cause.message ?: "Unknown persistence error")
        )
}
